File: shared/models/whois.py
import uuid
from datetime import datetime

from pydantic import BaseModel
from sqlalchemy import Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Field, SQLModel

from shared.enums.whois import WhoisLookupType
from shared.utils.datetime import utc_now


class WhoisRecord(SQLModel, table=True):
    __tablename__ = "whois_records"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True, index=True)

    query_value: str = Field(max_length=500, index=True, unique=True)
    lookup_type: WhoisLookupType
    queried_at: datetime = Field(
        default_factory=utc_now,
        description="When RDAP was last queried for this record",
    )

    handle: str = Field(default="", max_length=200)
    name: str = Field(default="", max_length=500, index=True)
    whois_server: str = Field(default="", max_length=200)
    object_class: str = Field(default="", max_length=100)
    rir: str = Field(default="", max_length=50)
    description: str = Field(default="", sa_type=Text)

    registration_date: datetime | None = None
    last_changed_date: datetime | None = None
    expiration_date: datetime | None = None

    registrant_name: str = Field(default="", max_length=500, index=True)
    registrant_email: str = Field(default="", max_length=500, index=True)
    registrar_name: str = Field(default="", max_length=500, index=True)
    abuse_email: str = Field(default="", max_length=500)

    nameservers: list | None = Field(default=None, sa_type=JSONB)
    domain_status: list | None = Field(default=None, sa_type=JSONB)
    dnssec: bool | None = None

    country: str = Field(default="", max_length=10, index=True)
    ip_version: int | None = None
    assignment_type: str = Field(default="", max_length=100)
    network_cidr: str = Field(default="", max_length=100, index=True)

    asn_range_start: int | None = None
    asn_range_end: int | None = None

    parsed_data: dict | None = Field(default=None, sa_type=JSONB)

    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)


class WhoisNameserver(SQLModel, table=True):
    __tablename__ = "whois_nameservers"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    whois_record_id: uuid.UUID = Field(
        foreign_key="whois_records.id",
        index=True,
        ondelete="CASCADE",
    )
    nameserver: str = Field(max_length=500, index=True)


class WhoisRecordRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    query_value: str
    lookup_type: str
    queried_at: datetime

    handle: str
    name: str
    whois_server: str
    object_class: str = Field(default="")
    rir: str
    description: str

    registration_date: datetime | None
    last_changed_date: datetime | None
    expiration_date: datetime | None

    registrant_name: str
    registrant_email: str
    registrar_name: str
    abuse_email: str

    nameservers: list | None
    domain_status: list | None
    dnssec: bool | None

    country: str
    ip_version: int | None
    assignment_type: str
    network_cidr: str

    asn_range_start: int | None
    asn_range_end: int | None

    parsed_data: dict | None

    created_at: datetime
    updated_at: datetime


class WhoisRecordSummary(BaseModel):
    id: uuid.UUID
    query_value: str
    lookup_type: str
    name: str
    registrant_name: str
    registrar_name: str
    country: str
    network_cidr: str
    registration_date: datetime | None
    expiration_date: datetime | None
    queried_at: datetime


class WhoisCorrelationResult(BaseModel):
    correlation_type: str
    correlation_value: str
    records: list[WhoisRecordSummary]
    count: int

File: frontend/src/lib/types/whois.ts
export type WhoisLookupType = 'DOMAIN' | 'IP' | 'ASN';

export interface WhoisEntityAddress {
	po_box: string;
	ext_address: string;
	street_address: string;
	locality: string;
	region: string;
	postal_code: string;
	country: string;
}

export interface WhoisEntity {
	handle: string;
	name: string;
	email: string;
	tel: string;
	type: string;
	url: string;
	rir: string;
	whois_server: string;
	address: WhoisEntityAddress | null;
}

export interface WhoisParsedData {
	lookup_type: WhoisLookupType;
	query: string;
	handle: string;
	parent_handle: string;
	name: string;
	whois_server: string;
	object_class: string;
	terms_of_service_url: string;
	copyright_notice: string;
	description: string[];
	registration_date: string | null;
	last_changed_date: string | null;
	expiration_date: string | null;
	rir: string;
	url: string;
	entities: {
		registrant: WhoisEntity[];
		administrative: WhoisEntity[];
		technical: WhoisEntity[];
		abuse: WhoisEntity[];
		billing: WhoisEntity[];
		registrar: WhoisEntity[];
		sponsor: WhoisEntity[];
		noc: WhoisEntity[];
		routing: WhoisEntity[];
	};
	unicode_name: string;
	nameservers: string[];
	status: string[];
	dnssec: boolean;
}

export interface WhoisRecordRead {
	id: string;
	query_value: string;
	lookup_type: string;
	queried_at: string;
	handle: string;
	name: string;
	whois_server: string;
	object_class: string;
	rir: string;
	description: string;
	registration_date: string | null;
	last_changed_date: string | null;
	expiration_date: string | null;
	registrant_name: string;
	registrant_email: string;
	registrar_name: string;
	abuse_email: string;
	nameservers: string[] | null;
	domain_status: string[] | null;
	dnssec: boolean | null;
	country: string;
	ip_version: number | null;
	assignment_type: string;
	network_cidr: string;
	asn_range_start: number | null;
	asn_range_end: number | null;
	parsed_data: WhoisParsedData | null;
	created_at: string;
	updated_at: string;
}

export interface WhoisRecordSummary {
	id: string;
	target_id: string | null;
	query_value: string;
	lookup_type: string;
	name: string;
	registrant_name: string;
	registrar_name: string;
	country: string;
	network_cidr: string;
	registration_date: string | null;
	expiration_date: string | null;
	queried_at: string;
}

export interface WhoisCorrelationResult {
	correlation_type: string;
	correlation_value: string;
	records: WhoisRecordSummary[];
	count: number;
}

export interface WhoisRefreshResponse {
	record: WhoisRecordRead;
	previous_queried_at: string | null;
}

export interface WhoisLookupRequest {
	query: string;
	store_in_db?: boolean;
}

export interface WhoisLookupResponse {
	record: WhoisRecordRead | null;
	data: Record<string, unknown>;
	cached: boolean;
}

export type WhoisEntityRole =
	| 'registrant'
	| 'administrative'
	| 'technical'
	| 'abuse'
	| 'billing'
	| 'registrar'
	| 'sponsor'
	| 'noc'
	| 'routing';

export const ENTITY_ROLE_LABELS: Record<WhoisEntityRole, string> = {
	registrant: 'Registrant',
	administrative: 'Administrative',
	technical: 'Technical',
	abuse: 'Abuse',
	billing: 'Billing',
	registrar: 'Registrar',
	sponsor: 'Sponsor',
	noc: 'NOC',
	routing: 'Routing'
};

export const CORRELATION_REASONS = [
	'registrant',
	'registrant_name',
	'registrar',
	'registrar_name',
	'nameserver',
	'network',
	'network_cidr'
] as const;
export type CorrelationReason = (typeof CORRELATION_REASONS)[number];

export const CORRELATION_REASON_LABELS: Record<
	CorrelationReason,
	{ full: string; match: string; short: string }
> = {
	registrant: { full: 'Registrant Name', match: 'registrant name', short: 'Registrant' },
	registrant_name: { full: 'Registrant Name', match: 'registrant name', short: 'Registrant' },
	registrar: { full: 'Registrar', match: 'registrar', short: 'Registrar' },
	registrar_name: { full: 'Registrar', match: 'registrar', short: 'Registrar' },
	nameserver: { full: 'Nameserver', match: 'nameserver', short: 'Nameserver' },
	network: { full: 'Network Block', match: 'network block', short: 'Network' },
	network_cidr: { full: 'Network Block', match: 'network block', short: 'Network' }
};

export type DomainStatusTone = 'pass' | 'warn' | 'fail' | 'info';

export interface DomainStatusInfo {
	label: string;
	tone: DomainStatusTone;
}

const DOMAIN_STATUS_RULES: ReadonlyArray<readonly [RegExp, DomainStatusInfo]> = [
	[/pending ?delete/i, { label: 'Pending delete', tone: 'fail' }],
	[/redemption/i, { label: 'Redemption period', tone: 'fail' }],
	[/hold/i, { label: 'On hold', tone: 'warn' }],
	[/inactive/i, { label: 'Inactive', tone: 'warn' }],
	[/transfer ?prohibited/i, { label: 'Transfer locked', tone: 'pass' }],
	[/delete ?prohibited/i, { label: 'Delete locked', tone: 'pass' }],
	[/update ?prohibited/i, { label: 'Update locked', tone: 'pass' }],
	[/renew ?prohibited/i, { label: 'Renew locked', tone: 'pass' }],
	[/auto ?renew/i, { label: 'Auto-renew period', tone: 'info' }],
	[/pending/i, { label: 'Pending change', tone: 'info' }],
	[/^(ok|active)$/i, { label: 'Active', tone: 'pass' }]
];

export function describeDomainStatus(code: string): DomainStatusInfo {
	const bare = code.replace(/\s+https?:\/\/\S+$/i, '').trim();
	for (const [re, info] of DOMAIN_STATUS_RULES) if (re.test(bare)) return info;
	return { label: bare, tone: 'info' };
}

export function isRedactedName(name: string | null | undefined): boolean {
	if (!name) return false;
	return /(redacted|privacy|proxy|withheld|not disclosed|data protected|gdpr|whoisguard|contact privacy)/i.test(
		name
	);
}
